src/zenml/materializers/default_materializer.py

```diff
@@ -0,0 +1,80 @@
+#  Copyright (c) ZenML GmbH 2023. All Rights Reserved.
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at:
+#
+#       https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+#  or implied. See the License for the specific language governing
+#  permissions and limitations under the License.
+"""Implementation of ZenML's default materializer.
+
+This is the materializer that is used if no other materializer is registered
+for a given data type.
+"""
+
+import os
+from typing import Any, Type
+
+import cloudpickle
+
+from zenml.enums import ArtifactType
+from zenml.io import fileio
+from zenml.logger import get_logger
+from zenml.materializers.base_materializer import BaseMaterializer
+
+logger = get_logger(__name__)
+
+DEFAULT_FILENAME = "artifact.pkl"
+
+
+class DefaultMaterializer(BaseMaterializer):
+    """Default materializer using cloudpickle.
+
+    This materializer is used if no other materializer is registered for a
+    given data type.
+    """
+
+    ASSOCIATED_TYPES = (object,)
+    ASSOCIATED_ARTIFACT_TYPE = ArtifactType.DATA
+
+    def load(self, data_type: Type[Any]) -> Any:
+        """Reads an artifact from a cloudpickle file.
+
+        Args:
+            data_type: The data type of the artifact.
+
+        Returns:
+            The loaded artifact data.
+        """
+        super().load(data_type)
+        filepath = os.path.join(self.uri, DEFAULT_FILENAME)
+        with fileio.open(filepath, "rb") as fid:
+            data = cloudpickle.load(fid)
```

Comment:
Does the error message when running this in a different python version than it was saved in give a reasonable explanation? If not, is it worth it to query our DB to find out the environment in which the artifact was saved and compare the python versions?

Reply:
Yeah makes sense, we should check this. I copied @strickvl's implementation that saves the python version to another file to check that. Do you think that's fine or would you prefer loading this info from the DB instead? How would you query the DB inside the materializer even? `list_artifacts(uri=...)[0]`?

Reply:
IMO the cleanest solution would be: All materializers get the `ArtifactResponseModel` passed in their `__init__` instead of just the URI. The pickle materializer stores the python version in the artifact metadata (useful anyway I guess?) and can then read that python version from there in `load()`. Wdyt?

Reply:
I agree that would be cleaner in this case, but IMO it's not worth changing the materializer interface over this (which would require adjusting the `__init__` of every materializer that implements it, deprecating the `uri` argument, adding corresponding backwards compatibility code, ...).

Generally, I'm also not sure whether coupling the materializer to the artifact is cleaner overall. Intuitively it seems good because each URI is currently assigned to a unique artifact, so we might as well pass the whole artifact. However, I do like the simplicity of the current interface, IMO it's very obvious what it's supposed to do without requiring any other classes for context.

Reply:
Depends, I guess we might want to make it clean at some point right? Unfortunately it doesn't even make sense to have it in the `__init__` as a materializer that just gets initialized to save some data does not have the response model available at that time I guess.

Ideally I think we would have `classmethods` for both save/load, where the `load` method would get the model. But you'd say this PR is not the time to adjust that right?

Reply:
Yeah I would not do it now since I'm not even convinced that coupling materializers to artifact classes is necessarily cleaner.